### PR TITLE
Add toBiMap() for RichIterable. Fixes #306

### LINKED ISSUE #306: Implement toBiMap on RichIterable
There is currently toMap and toSortedMap but there is no toBiMap.  The signature should be as follows:

```
<NK, NV> MutableBiMap<NK, NV> toBiMap(
        Function<? super T, ? extends NK> keyFunction,
        Function<? super T, ? extends NV> valueFunction);

```

### PR DESCRIPTION
**Pull request by sendilkumarn**
Fixes #306 

cc: @donraab

### CODE REVIEW

**Comment by motlin** [comment]
What is the behavior of MapCollectProcedure when there are duplicates in
the BiMap?

Could you add a test where you map an iterable of strings to their first
character, and there are duplicates in either the mapped keys, values, or
both?

**Comment by sendilkumarn** [comment]
@motlin I am not sure I understood it correctly. Can you please elaborate? (P.S. I am still a newbie here)

**Comment by motlin** [comment]
This is new behavior so I think it’s just something we should address in
the pull request before merging. You’ll probably have to replace
MapCollectProcedure with the desired behavior.

**Comment by sendilkumarn** [comment]
@motlin Won't it affect the other types?  Since every iterable is using the same mapCollectProcedure.  Do you mean implement something like `BiMapCollectProcedure`?

Also running through the values/ keys while inserting will make them slower right?

**Comment by motlin** [comment]
It shouldn't have a noticeable performance impact since we can use an
atomic operation. If we don't care much about nulls, we can use similar
logic to groupByUniqueKey.

if (this.map.put(key, object) != null)
{
    throw new IllegalStateException("Key " + key + " already exists in map!");
}

**Comment by motlin** [comment]
Also, AbstractMutableBiMap#put is already incredibly inefficient. So I
wouldn't worry too much about making it worse.

**Comment by donraab** [comment]
@motlin Is it enough to just call put for BiMap?  I thought you would have to check to see if the value was there for another key first?

**Comment by nikhilnanivadekar** [comment]
@donraab put on BiMap takes cares of verifying if a key or a value exists, before adding to Map and Inverse

**Comment by donraab** [comment]
@nikhilnanivadekar ok, cool!

**Comment by motlin** [comment]
To add to what Nikhil said, MutableBiMap.put(newKey, existingValue) will
throw. We should throw, but may want to throw a more helpful error message.

MutableBiMap.put(existingKey, newValue) will return the old key, which we
can detect is non-null and throw.

My point about efficiency is that we're so inefficient, I wouldn't mind
just calling containsKey and containsValue if the code looks simpler.

**Comment by sendilkumarn** [comment]
> To add to what Nikhil said, MutableBiMap.put(newKey, existingValue) will
throw. We should throw, but may want to throw a more helpful error message.

This is happening currently, Maybe we can format the error message slightly here.

> MutableBiMap.put(existingKey, newValue) will return the old key, which we
can detect is non-null and throw.

This is really an interesting case. I can say there are three scenarios here.

1 ->  newValue is new and does not exist in inverse. Then it is a normal update operation. We should allow this. 

2 -> newValue is already a key in inverse. In that case, throwing an error is logically correct.

3 -> newValue is a key in the map. We can allow that too. Since it will not complicate the already bounded bindings. 

Is the understanding here correct? Or am I missing some context?

**Comment by motlin** [comment]
> 1 -> newValue is new and does not exist in inverse. Then it is a normal
update operation. We should allow this.
I'm saying we should specifically not allow this. It's confusing, and order
sensitive. It's almost definitely a bug. If we allow the iteration pattern
to complete, the state is already corrupted, and it can be hard to figure
out after the fact why some map pairs are missing.

> 2 -> newValue is already a key in inverse. In that case, throwing an
error is logically correct.
Sure, but a BiMap is symmetrical. If you believe that bimap.put(k, v) ought
to throw, wouldn't you agree that map.inverse().put(v, k) ought to throw?

> 3 -> newValue is a key in the map. We can allow that too. Since it will
not complicate the already bounded bindings.
You mean if K and V are the same type? The keys and values are logically
two separate sets. It doesn't matter if the same element is in both sets.

**Comment by sendilkumarn** [comment]
> I'm saying we should specifically not allow this. It's confusing, and order
sensitive. It's almost definitely a bug. 

Adding this logic in the `put` method of the `biMap` will make the biMap not updatable.
OTOH, handling this in the `(bi)mapCollectProecedure` will make biMap not so consistent. 

> If we allow the iteration pattern
to complete, the state is already corrupted, and it can be hard to figure
out after the fact why some map pairs are missing.

I am more aligned towards the explicitness in the `BiMap` at least for the `put` should behave like a normal `map`.  IMO it feels natural. Yes there will be some `hard to figure out` stuffs here. 

May be we can introduce something like `strictBiMap` that adds this constraint 

> wouldn't you agree that map.inverse().put(v, k) ought to throw?

iff k is present as a `key` in the (not-inverse) map.

**Comment by motlin** [comment]
I wasn't suggesting changing the behavior of put(), just a thought
experiment.

Let's consider collect(). Let's say we take the two functions passed to
toBiMap() and instead pass them to collect().

keys = richIterable.collect(keyFunction)
values = richIterable.collect(valueFunction)

Say keys are all unique, but values contain duplicates. If we follow your
points (1) and (2), then the behavior regarding duplicates is not
symmetrical.

richIterable.toMap(keyFunction, valueFunction) throws
richIterable.toMap(valueFunction, keyFunction).inverse() doesn't throw

My assertion is that these two lines must have the same behavior. They must
return the same value or both throw. If we agree that either must throw,
then both must throw.

**Comment by sendilkumarn** [comment]
@motlin IIUC, In the `BiMapCollectProcedure`  before we put values into the map. We will check if all the available keys and values are unique w.r.t keys and values iterable?

If they are unique ✅, else throw an error.

**Comment by motlin** [comment]
That sounds right. And to get inspiration for what the error message might
say, look to groupByUniqueKey

**Comment by nikhilnanivadekar** [comment]
I would say iterate just once though to check if there are duplicates and to put, don’t do a precheck.

**Comment by motlin** [comment]
That doesn’t work for LazyIterables

**Comment by sendilkumarn** [comment]
@motlin can you have a look at it now?

File: eclipse-collections-code-generator/src/main/resources/impl/map/immutable/immutablePrimitiveObjectEmptyMap.stg [hunk 1]
```diff
@@ -584,6 +586,12 @@ final class Immutable<name>ObjectEmptyMap\<V> implements Immutable<name>ObjectMa
         return SortedMaps.mutable.with();
     }
 
+    @Override
+    public \<NK, NV> MutableBiMap\<NK, NV> toBiMap(Function\<? super V, ? extends NK> keyFunction, Function\<? super V, ? extends NV> valueFunction)
+    {
+        return BiMaps.mutable.with();
```
[the remark this answers is not in the document]
Comment (sendilkumarn) [reply eclipse-collections-code-generator/src/main/resources/impl/map/immutable/immutablePrimitiveObjectEmptyMap.stg]: I think that is the custom followed here and `with()` in turn calls `empty()`

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/bimap/mutable/HashBiMap.java [hunk 2]
```diff
@@ -70,6 +73,15 @@ public HashBiMap(Map<K, V> map)
...
     }
 
+    public <E> MutableBiMap<K, V> collectKeysAndValues(
+            Iterable<E> iterable,
+            Function<? super E, ? extends K> keyFunction,
+            Function<? super E, ? extends V> valueFunction)
+    {
+        Iterate.forEach(iterable, new BiMapCollectProcedure<>(this, keyFunction, valueFunction));
+        return this;
+    }
```
[the remark this answers is not in the document]
Comment (sendilkumarn) [reply eclipse-collections/src/main/java/org/eclipse/collections/impl/bimap/mutable/HashBiMap.java]: This method was missing but it is not an override either.

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/block/procedure/BiMapCollectProcedure.java [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+{
+    private static final long serialVersionUID = 1L;
+
+    private final MutableBiMap<K, V> biMap;
+    private final Function<? super T, ? extends K> keyFunction;
+    private final Function<? super T, ? extends V> valueFunction;
+
+    public BiMapCollectProcedure(MutableBiMap<K, V> newMap, Function<? super T, ? extends K> newKeyFunction)
+    {
+        this(newMap, newKeyFunction, (Function<T, V>) Functions.identity());
```
[the remark this answers is not in the document]
Comment (sendilkumarn) [reply eclipse-collections/src/main/java/org/eclipse/collections/impl/block/procedure/BiMapCollectProcedure.java]: removed

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/block/procedure/BiMapCollectProcedure.java [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+    @Override
+    public void value(T object)
+    {
+        K key = this.keyFunction.valueOf(object);
+        V value = this.valueFunction.valueOf(object);
+
+        if (this.biMap.containsKey(key) && this.biMap.inverse().containsKey(value))
+        {
+            throw new IllegalArgumentException("The Key " + key + " and value " + value + " is already present.");
```
[the remark this answers is not in the document]
Comment (sendilkumarn) [reply eclipse-collections/src/main/java/org/eclipse/collections/impl/block/procedure/BiMapCollectProcedure.java]: >  where all keys and values are unique.

No there are tests added to check when the key and value are not unique. Can you be more clear on what you are trying to tell?

File: eclipse-collections-code-generator/src/main/resources/impl/map/immutable/immutablePrimitiveObjectEmptyMap.stg (hunk 1 above)
[the remark this answers is not in the document]
Comment (nikhilnanivadekar) [reply eclipse-collections-code-generator/src/main/resources/impl/map/immutable/immutablePrimitiveObjectEmptyMap.stg]: we delegate with() to empty(). empty() came after with() and we kept with() around to not break users. Since empty() came we always use empty() as a convention.

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/block/procedure/BiMapCollectProcedure.java [hunk 5]
```diff
@@ -0,0 +1,55 @@
+/*
+ * Copyright (c) 2018 Goldman Sachs and others.
```
[the remark this answers is not in the document]
Comment (nikhilnanivadekar) [reply eclipse-collections/src/main/java/org/eclipse/collections/impl/block/procedure/BiMapCollectProcedure.java]: @sendilkumarn the copyright should conform to the details given here: https://www.eclipse.org/projects/handbook/#ip-copyright-headers

So, in this case, you need to put Copyright (c) <Enter the organization on behalf of which you signed the Eclipse Contributor Agreement>. and then rest of the paragraph is same.

File: unit-tests/src/test/java/org/eclipse/collections/impl/AbstractRichIterableTestCase.java [hunk 6]
```diff
@@ -1363,6 +1364,14 @@ public void toSortedMapBy()
         Verify.assertListsEqual(Lists.mutable.with(3, 2, 1), map.keySet().toList());
     }
 
+    @Test
+    public void toBiMap()
+    {
+        RichIterable<Integer> integers = this.newWith(1, 2, 3);
+        MutableBiMap<String, String> biMap = integers.toBiMap(Object::toString, Object::toString);
+        Assert.assertEquals(UnifiedMap.newWithKeysValues("1", "1", "2", "2", "3", "3"), biMap);
```
Comment (motlin) [inline unit-tests/src/test/java/org/eclipse/collections/impl/AbstractRichIterableTestCase.java]: This is what I mean by all elements are mapped to unique keys and values. You should be able to change the key function, value function, or both functions to one that returns a duplicate element (like `ignored -> "constant"`) and see that the behavior is incorrect.

File: .idea/vcs.xml [hunk 7]
```diff
@@ -2,5 +2,6 @@
 <project version="4">
   <component name="VcsDirectoryMappings">
     <mapping directory="$PROJECT_DIR$" vcs="Git" />
+    <mapping directory="$PROJECT_DIR$/envoy" vcs="Git" />
```
Comment (motlin) [inline .idea/vcs.xml]: Please revert the change to vcs.xml

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/block/procedure/BiMapCollectProcedure.java [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+    {
+        K key = this.keyFunction.valueOf(object);
+        V value = this.valueFunction.valueOf(object);
+
+        if (this.biMap.containsKey(key) && this.biMap.inverse().containsKey(value))
+        {
+            throw new IllegalArgumentException("The Key " + key + " and value " + value + " is already present.");
+        }
+
+        this.biMap.put(this.keyFunction.valueOf(object), this.valueFunction.valueOf(object));
```
Comment (motlin) [inline eclipse-collections/src/main/java/org/eclipse/collections/impl/block/procedure/BiMapCollectProcedure.java]: Why call the functions again? Why not:
```java
        this.biMap.put(key, value);
```

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/bimap/mutable/HashBiMap.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (sendilkumarn) [reply eclipse-collections/src/main/java/org/eclipse/collections/impl/bimap/mutable/HashBiMap.java]: I will need it here https://github.com/eclipse/eclipse-collections/blob/4e177372405a7956bcc6a49c6281e72ce0e229ae/eclipse-collections/src/main/java/org/eclipse/collections/impl/collection/mutable/AbstractCollectionAdapter.java#L842

File: unit-tests/src/test/java/org/eclipse/collections/impl/AbstractRichIterableTestCase.java (hunk 6 above)
Comment (sendilkumarn) [reply unit-tests/src/test/java/org/eclipse/collections/impl/AbstractRichIterableTestCase.java]: Have you checked it here https://github.com/eclipse/eclipse-collections/blob/4e177372405a7956bcc6a49c6281e72ce0e229ae/unit-tests/src/test/java/org/eclipse/collections/impl/AbstractRichIterableTestCase.java#L1375-L1387

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/block/procedure/BiMapCollectProcedure.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (sendilkumarn) [reply eclipse-collections/src/main/java/org/eclipse/collections/impl/block/procedure/BiMapCollectProcedure.java]: https://github.com/eclipse/eclipse-collections/pull/667#discussion_r244649132

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/bimap/mutable/HashBiMap.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (motlin) [reply eclipse-collections/src/main/java/org/eclipse/collections/impl/bimap/mutable/HashBiMap.java]: Couldn't you inline the implementation into AbstractCollectionAdapter and delete the method?

File: unit-tests/src/test/java/org/eclipse/collections/impl/AbstractRichIterableTestCase.java (hunk 6 above)
Comment (motlin) [reply unit-tests/src/test/java/org/eclipse/collections/impl/AbstractRichIterableTestCase.java]: These test two of the three scenarios I mentioned, and I think the one that's still missing is the important one.

Comment (sendilkumarn) [reply unit-tests/src/test/java/org/eclipse/collections/impl/AbstractRichIterableTestCase.java]: You mean test with duplicate keys? Added that. 👍

Comment (motlin) [reply unit-tests/src/test/java/org/eclipse/collections/impl/AbstractRichIterableTestCase.java]: Yes! And I see it uncovered the bug and you've fixed it. I hope that wasn't too frustrating but I think it's more rewarding to add tests until you can find the one that's red and get it to go to green. I'm going to add a few more suggestions but we're down to lower priority style stuff and the overall change is in good shape.

Comment (motlin) [reply unit-tests/src/test/java/org/eclipse/collections/impl/AbstractRichIterableTestCase.java]: I made the changes in my fork and created a PR from mine to yours @sendilkumarn 

https://github.com/sendilkumarn/eclipse-collections/pull/1

You don't have to actually merge it. If you like the changes you could squash it for example. I made the PR to make it visible.

**Comment by donraab** [comment]
@sendilkumarn @motlin Is there anything outstanding on this PR or is it ready for final review and approval?

**Comment by motlin** [comment]
I’d suggest one last cleanup just to commit messages. Both commits should have a title like “Implement RichIterable.toBiMap()” without the issue number #1. The number corresponds with your issue tracker. Here it links to our initial import.

**Comment by sendilkumarn** [comment]
Is this open for any reason? Can we merge this ?

**Comment by donraab** [comment]
The build was failing.  I restarted it.  Please rebase and I will approve and merge once the build passes.

**Comment by nikhilnanivadekar** [comment]
All the builds have gone through and @motlin has approved the review. Please rebase on to master and also squash the commits. Please update the commit message to include the word "Fixes" so that #306 is auto-closed. 
Thanks for your patience and contribution!

**Comment by nikhilnanivadekar** [comment]
@sendilkumarn can you please rebase on to master and fix the commit messages, squash the commits. Thanks!

**Comment by sendilkumarn** [comment]
Sure I will merge it with upstream.

**Comment by donraab** [comment]
@sendilkumarn Thank you for the contribution!
